core/src/kvstore/io_utils.rs

```diff
@@ -25,12 +30,126 @@ pub struct SharedWriter {
     pos: u64,
 }
 
+#[derive(Debug)]
+pub struct CRCWriter<W: Write> {
+    wtr: W,
+    buf: Vec<u8>,
+    pos: usize,
+    cap: usize,
+}
+
+#[derive(Debug)]
+pub struct CRCReader<R: Read> {
+    rdr: R,
+    buf: Vec<u8>,
+    pos: usize,
+    chunk_size: usize,
+}
+
+impl<W: Write> CRCWriter<W> {
+    pub fn new(inner: W, chunk_size: usize) -> CRCWriter<W> {
+        if chunk_size <= 8 {
+            panic!("chunk_size must be > 8");
+        }
+
+        CRCWriter {
+            wtr: inner,
+            buf: vec![0; chunk_size],
+            pos: 0,
+            cap: chunk_size - 8,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn into_inner(mut self) -> io::Result<W> {
+        self.flush()?;
+        Ok(self.wtr)
+    }
+
+    pub fn get_ref(&self) -> &W {
+        &self.wtr
+    }
+
+    pub fn get_mut(&mut self) -> &mut W {
+        &mut self.wtr
+    }
+}
+
+impl<R: Read> CRCReader<R> {
+    pub fn new(inner: R, chunk_size: usize) -> CRCReader<R> {
+        if chunk_size <= 8 {
+            panic!("chunk_size must be > 8");
+        }
+
+        CRCReader {
+            rdr: inner,
+            buf: vec![0; chunk_size - 8],
+            pos: chunk_size,
+            chunk_size,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn into_inner(self) -> R {
+        self.rdr
+    }
+}
+
 impl SharedWriter {
     pub fn new(buf: Arc<RwLock<Vec<u8>>>) -> SharedWriter {
         SharedWriter { buf, pos: 0 }
     }
 }
 
+impl<R> CRCReader<R>
```

Comment:
how is this block different from the block beginning at line 78?

Reply:
It's not, I just wanted to keep `pub` types, functions, methods, etc. to the top of a file for easy skimming.